docs: introduce ADR on slashing on the provider chain (#1252)
cosmos/interchain-security: docs/docs/adrs/adr-013-equivocation-slashing.md

```protobuf
type DuplicateVoteEvidence struct {
	VoteA *Vote `json:"vote_a"`
	VoteB *Vote `json:"vote_b"`

	// abci specific information
	TotalVotingPower int64
	ValidatorPower   int64
	Timestamp        time.Time
}
```
The "abci specific information" fields cannot be trusted because they are not signed. Therefore,
we can use neither `ValidatorPower` for slashing on the provider chain, nor the `Timestamp` to check the evidence age. We can get the `infractionHeight` from the votes, but this `infractionHeight` corresponds to the infraction height on the consumer and **not** on the provider chain.

## Proposed solution
As a first iteration, we propose the following approach. At the moment the provider receives evidence of equivocation on a consumer:
1. slash all the undelegations and redelegations using `slashFactor`;
2. slash all delegations using as voting `power` the sum of the voting power of the misbehaving validator and the power of all the ongoing undelegations and redelegations.

> **Evidence expiration:** Additionally, because we cannot infer the actual time of the evidence (i.e., the timestamp of the evidence cannot be trusted), we do not consider _evidence expiration_ and hence old evidence is never ignored (e.g., the provider would act on 3 year-old evidence of equivocation on a consumer).

Reviewer: I would add something in here about how we will not slash validators if they have reassigned their consumer key more than one unbonding period ago

Author: Added a paragraph on this.

---
sidebar_position: 14

> title: ADR Template

Reviewer 2: Also please fix the title of ADR 11

#### Slashing undelegations and redelegations
To slash undelegations, `Slash` goes through all undelegations and checks whether they started before or after the infraction occurred. If an undelegation started before the `infractionHeight`, then it is **not** slashed, otherwise it is slashed by `slashFactor`.

The slashing of redelegations happens in a similar way, meaning that `Slash` goes through all redelegations and checks whether the redelegations started before or after the `infractionHeight`.

#### Slashing delegations
Besides undelegations and redelegations, the validator's delegations need to also be slashed.
This is performed by deducting the appropriate amount of tokens from the validator. Note that this deduction is computed based on the voting `power` the misbehaving validator had at the height of the equivocation. As a result of the tokens deduction, 
the [tokens per share](https://docs.cosmos.network/v0.47/modules/staking#delegator-shares)
reduce and hence later on, when delegators undelegate or redelegate, the delegators retrieve back less
tokens, effectively having their tokens slashed. This approach of slashing delegations does not utilize the

> `infractionHeight` in any way and hence the following scenario could occur:

Reviewer 3: This behaviour is standard in the SDK, right? If so, could you make that more explicit, maybe link a place where this behaviour is stated? It seems like you are describing a flaw in the SDK, so it would be good to have more context, preferably just a link to the SDK docs about this

Author: Yes, this is standard SDK behaviour. I'm not aware of any SDK documentation clearly specifying this behaviour or such a scenario. The best I could find is the link already provided in regards to [delegator shares](https://docs.cosmos.network/v0.47/modules/staking#delegator-shares) where it's written:

> The reason for this mechanism is to simplify the accounting around slashing. Rather than iteratively slashing the tokens of every delegation entry, instead the Validators total bonded tokens can be slashed, effectively reducing the value of each issued delegator share.

Reviewer 3: Thanks, that works for me!

## Proposed solution
As a first iteration, we propose the following approach. At the moment the provider receives evidence of equivocation on a consumer:
1. slash all the undelegations and redelegations using `slashFactor`;
2. slash all delegations using as voting `power` the sum of the voting power of the misbehaving validator and the power of all the ongoing undelegations and redelegations.

**Evidence expiration:** Additionally, because we cannot infer the actual time of the evidence (i.e., the timestamp of the evidence cannot be trusted), we do not consider _evidence expiration_ and hence old evidence is never ignored (e.g., the provider would act on 3 year-old evidence of equivocation on a consumer).
Additionally, we do not need to store equivocation evidence to avoid slashing a validator more than once, because we [do not slash](https://github.com/cosmos/cosmos-sdk/blob/v0.47.0/x/evidence/keeper/infraction.go#L94) tombstoned validators and we [tombstone](https://github.com/cosmos/cosmos-sdk/blob/v0.47.0/x/evidence/keeper/infraction.go#L138) a validator when slashed.

> We do not act on evidence that was signed by a validator [consensus key](https://tutorials.cosmos.network/tutorials/9-path-to-prod/3-keys.html#what-validator-keys) that is _pruned_ when we receive the evidence. We prune a validator's consensus key if the validator has assigned a new consumer key (using `MsgAssignConsumerKey`) and an unbonding period on the consumer chain has elapsed (see [key assignment ADR](https://github.com/cosmos/interchain-security/blob/main/docs/docs/adrs/adr-001-key-assignment.md)). Note that the provider chain is informed that the unbonding period has elapsed on the consumer when the provider receives a `VSCMaturedPacket` and because of this, if the consumer delays the sending of a `VSCMaturedPacket`, we would delay the pruning of the key as well.

Reviewer 3: Could you mention here that a malicious consumer chain can not delay sending VSCMaturedPackets forever? (This should hold true due to the VSCTimeout) I know this is mentioned in the ICS protocol, but might be good to have it here too for context

Author: I believe a malicious consumer chain can delay a `VSCMaturedPacket` forever. If the consumer chain is malicious, it could never write the  `VSCMaturedPacket` packet in its store for the packet to be sent. In my view, the timeout assumption holds if the consumer chain tries to actually send the packet and this is not received on the other end.

###  Slashing on the provider
We see that in the single-chain slashing case, we use the `infractionHeight` and the voting `power` to be able to slash.
In order to slash the provider for a consumer equivocation we need to have the provider's `infractionHeight`
and voting `power`. However, we do **not** have those values. We only have the `infractionHeight` in the consumer chain
(that we can extract from the equivocated votes), but we do not know to what provider height this `infractionHeight` corresponds.
Unless we have a way to find the corresponding `infractionHeight` and `power` in the provider chain, we cannot slash the provider in the same way as we would slash in the single-chain case.

Someone might think that the problem of figuring out the corresponding `infractionHeight` and `power` values in the provider chain is easy because we could have the consumer chain send us this information.
However, we consider that the application on the consumer chain could be _malicious_ and hence we  cannot really trust anything that stems from the _application state_ of the consumer chain.

> Note that when a relayer or a user sends evidence through a [MsgSubmitConsumerDoubleVoting](https://github.com/cosmos/interchain-security/pull/1232/files) message, we get what is contained in the [DuplicateVoteEvidence](https://github.com/cometbft/cometbft/blob/v0.37.0/types/evidence.go#L35):

Author: Resolving this for now and leaving **un**resolved your comment [here](https://github.com/cosmos/interchain-security/pull/1252#discussion_r1317286951) so we have a single point of discussion in regards to light client attacks.

	// abci specific information
	TotalVotingPower int64
	ValidatorPower   int64
	Timestamp        time.Time
}
```
The "abci specific information" fields cannot be trusted because they are not signed. Therefore,
we cannot use `ValidatorPower` in any way for slashing in the provider chain. We can get the `infractionHeight` from the votes, but as already mentioned, this `infractionHeight` corresponds to the infraction height on the consumer and **not** on the provider chain.
Furthermore, note that the [Timestamp](https://github.com/cometbft/cometbft/blob/v0.37.0/types/vote.go#L55) in the votes is just the [BFT time](https://github.com/cometbft/cometbft/blob/v0.37.0/spec/consensus/bft-time.md), and the time could be anything since a misbehaving validator could have included any time in the votes.

Finally, note that in the single-chain case, we trust the underlying consensus engine (CometBFT) and hence when we receive evidence from CometBFT we can act on it. In the case of provider and consumer chains however we cannot trust the evidence as is (e.g., evidence is submitted by a relayer that could set arbitrary values to the "abci specific information" fields).

> #### Clock drift between the provider and the consumer chain

Reviewer 2: I would remove this section from here. For now let's have the discussion on the issue or PR.

> #### Clock drift between the provider and the consumer chain

Reviewer 2: I would remove this section from here. For now let's have the discussion on the issue or PR.

IBC clients contain a `maxClockDrift` parameter but this is [only used](https://github.com/cometbft/cometbft/blob/v0.37.0/light/verifier.go#L176)
to reject headers that come slightly from the future. For example, assume a chain `A` that has a light client `lc` that tracks chain `B` and that has `maxClockDrift` of 10 seconds. If chain `A` has time `12:58:31` and then `lc` receives a header with time `12:58:42` (that is 11 seconds from the future), then `lc` would reject his client update.
However, even with this `maxClockDrift` constraint we still do not know the time at the chain `B`, e.g., `B`'s BFT time could be 5 years ahead compared to chain's `A` BFT time.

(TODO: add more on the `trustingPeriod`) only tells us from the provider side that we have received a message from the other chain in the last `trustingPeriod` (2 weeks in Cosmos Hub) before ... so doesn't add much either ...

We currently investigate on whether we can bound the clock drift between the provider and the consumer chain.

## Proposed solution

> As a first iteration, in order to implement the slashing functionality as soon as possible, we propose an aggressive (i.e., would slash more delegators than in the single-chain case) slashing  approach because this allows for a straight-forward implementation.
> In this approach, at the moment we receive evidence, at _evidence height_, we:

Reviewer 2: ```suggestion
As a first iteration, we propose the following approach. At the moment the provider receive evidence of equivocation on a consumer:
```

1. slash all the undelegations and redelegations using `slashFactor`;
2. slash all delegations using as voting `power` the sum of the voting power of the misbehaving validator and the power of all the ongoing undelegations and redelegations.

> **Evidence expiration:** Additionally, in this approach and because we cannot infer the actual time of the evidence (i.e., timestamp in evidence could contain anything), we do not consider _evidence expiration_ and hence old evidence is never ignored (e.g., we would act on 3 year-old evidence).

Reviewer 2: ```suggestion
**Evidence expiration:** Additionally, because we cannot infer the actual time of the evidence (i.e., the timestamp of the evidence cannot be trusted), we do not consider _evidence expiration_ and hence old evidence is never ignored (e.g., the provider would act on 3 year-old evidence of equivocation on a consumer).
```

### Implementation

> The aggressive approach allows for multiple simplifications. Specifically, we intend to introduce the following logic in the [HandleConsumerDoubleVoting](https://github.com/cosmos/interchain-security/pull/1232) method:

Reviewer 2: ```suggestion
The following logic needs to be added to the [HandleConsumerDoubleVoting](https://github.com/cosmos/interchain-security/pull/1232) method:
```

Reviewer 2: What about light client attacks? Do we need to differentiate?

Author: The described approach (i.e,. adding code `HandleConsumerDoubleVoting`) won't be useful for light client attacks as is. To slash for light client attacks we would need to do the same in  `HandleConsumerMisbehaviour` in response to `MsgSubmitConsumerMisbehaviour` messages.

However, I was thinking of this ADR specifically for double voting misbehaviour and how to slash in those cases.
It's not clear to me exactly how we verify (and punish for) light client attacks today since the staking module [Slash](https://github.com/cosmos/cosmos-sdk/blob/main/x/staking/keeper/slash.go#L193) is for equivocations. I would need to investigate this more in order to answer your questions.

```go
undelegationsTotalAmount := sdk.NewInt(0)
for _, v := range k.stakingKeeper.GetUnbondingDelegationsFromValidator(ctx, validatorAddress) {
    for _, entry := range v.Entries {
        undelegationsTotalAmount = undelegationsTotalAmount.Add(entry.InitialBalance)
    }
}

redelegationsTotalAmount := sdk.NewInt(0)
for _, v := range k.stakingKeeper.GetRedelegationsFromSrcValidator(ctx, validatorAddress) {
    for _, entry := range v.Entries {
        redelegationsTotalAmount = redelegationsTotalAmount.Add(entry.InitialBalance)
    }
}

infractionHeight := 0

> totalPower := validators voting Power + undelegationsTotalAmount.Add(redelegationsTotalAmount)

Reviewer 2: `undelegationsTotalAmount` and `redelegationsTotalAmount` are tokens, while `totalPower` is power.

Author: Good catch. Introduced [TokensToConsensusPower](https://github.com/cosmos/cosmos-sdk/blob/v0.47.0/x/staking/keeper/power_reduction.go) to convert `undelegationsTotalAmount.Add(redelegationsTotalAmount)` to power that we can then add to the validator's `power`.

slashFraction := k.slashingKeeper.SlashFractionDoubleSign(ctx)

k.stakingKeeper.Slash(ctx, validatorConsAddress, infractionHeight, totalPower, slashFraction, DoubleSign)
```

**Infraction height:** We provide a zero `infractionHeight` to the [Slash](https://github.com/cosmos/cosmos-sdk/blob/v0.47.0/x/staking/keeper/slash.go#L33) method in order to slash all ongoing undelegations and redelegations (see checks in [Slash](https://github.com/cosmos/cosmos-sdk/blob/v0.47.0/x/staking/keeper/slash.go#L92), [SlashUnbondingDelegation](https://github.com/cosmos/cosmos-sdk/blob/v0.47.0/x/staking/keeper/slash.go#L195), and [SlashRedelegation](https://github.com/cosmos/cosmos-sdk/blob/v0.47.0/x/staking/keeper/slash.go#L249)).

> **Power:** We pass the sum of the voting power of the misbehaving validator and the power of all the ongoing undelegations and redelegations. This is a slightly more aggressive approach than just providing the voting `power` at the evidence height. If we assume that the `slashFactor` is 5%, then the `power` we pass is `validatorPower + totalPower(undelegations) + totalPower(redelegations)`.

Reviewer 2: > This is a slightly more aggressive approach than just providing the voting `power` at the evidence height.

What do you mean by "voting `power` at the evidence height"? Is the power when the evidence was received? Isn't that the same as "the voting power of the misbehaving validator" from above?

Author: Wanted to write "the voting `power` at the infraction height." instead. This is not necessarily the case so I removed the sentence altogether to avoid confusion.

> Hence, when the `Slash` method slashes all the undelegations and redelegations it would end up with `0.05 * validatorPower + 0.05 * totalPower(undelegations) + 0.05 * totalPower(redelegations) - 0.05 * totalPower(undelegations) - 0.05 * totalPower(redelegations) = 0.05 * validatorPower` and hence it would slash 5% of the `validatorPower` at evidence height.

Reviewer 2: ```suggestion
Hence, when the `Slash` method slashes all the undelegations and redelegations it would end up with `0.05 * power - 0.05 * totalPower(undelegations) - 0.05 * totalPower(redelegations) = 0.05 * validatorPower` and hence it would slash 5% of the validator's power at the height when the evidence is received.
```

> **Storing evidence:** As mentioned, we do not expire evidence with this aggressive approach. However, because we do not want to slash the same validator more than once for the same infraction, we have to store the evidence in some _cache_ in order to check if we have already slashed that validator. We intend to do the following:

Reviewer 2: We don't need to store evidence. Validators that are slashed for double signing are tombstoned. Tombstoned validators cannot be slashed again.

Author: Yes, missed this. 🤦 
Removed the section altogether and added the following sentence:
>Note that we do not need to store equivocation evidence to avoid slashing a validator more than once, because we [do not slash](https://github.com/cosmos/cosmos-sdk/blob/v0.47.0/x/evidence/keeper/infraction.go#L94) tombstoned validators and we [tombstone](https://github.com/cosmos/cosmos-sdk/blob/v0.47.0/x/evidence/keeper/infraction.go#L138) a validator when slashed.

```go
// hashEvidence should only use the signed parts of the evidence to generate the hash
// therefore, it should only use VoteA and VoteB to avoid someone changing other parts of DuplicateVoteEvidence
// and resubmitting
key := hashEvidence(dve DuplicateVoteEvidence) 

if key not in cache {
	store (key, validatorAddr) in cache
	perform slashing
} else {
	// do not slash
}
```
To prevent this cache from growing arbitrarily big, we introduce additional checks. If the evidence  is for a validator that is not in the active validator set we skip from storing it in the cache, etc.
We also periodically clean up the cache by checking if the validator `validatorAddr` of a cache entry  is still in the active set and if not we remove that entry from the cache.

### Positive
With the proposed approach we can quickly implement slashing functionality on the provider chain for consumer chain equivocations.
This approach does not need to change the staking module and therefore does not change in any way how slashing is performed today for a single chain.

### Negative

> We _definitely_ slash more when it comes to undelegations and redelegations because we slash for all of them without considering an `infractionHeight`.
> We _potentially_ slash more than what we would have slashed if we knew the voting `power` at the corresponding `infractionHeight` in the provider chain.

Reviewer 2: ```suggestion

- We _definitely_ slash more when it comes to undelegations and redelegations because we slash for all of them without considering an `infractionHeight`.
- We _potentially_ slash more than what we would have slashed if we knew the voting `power` at the corresponding `infractionHeight` in the provider chain.
```

Reviewer 2: Also add a bullet point about the evidence age.

> ## Can we expire the evidence?

Reviewer 2: Remove for now and we have the discussion on the issue.

This is _orthogonal_ to the above solution and describes a way to check whether evidence has expired or not.

(TODO: Add Josef's idea on using Lamport clocks to expire old evidence.)

## References

> * [feat: add handler for consumer double voting #1232](https://github.com/cosmos/interchain-security/pull/1232)

Reviewer 2: ```suggestion
* [ADR 005: Cryptographic verification of equivocation evidence](https://github.com/cosmos/interchain-security/blob/main/docs/docs/adrs/adr-005-cryptographic-equivocation-verification.md)
* [EPIC tracking cryptographic equivocation feature](https://github.com/cosmos/interchain-security/issues/732)
```

> * [Cryptographic equivocation slashing design](https://forum.cosmos.network/t/cryptographic-equivocation-slashing-design/11400)

Reviewer 2: ```suggestion
* [Cosmos Hub Forum discussion on cryptographic equivocation slashing](https://forum.cosmos.network/t/cryptographic-equivocation-slashing-design/11400)
```

---
sidebar_position: 14
title: ADR Template
---
# ADR 013: Slashing on the provider for consumer equivocation

## Changelog
* 1st Sept. 2023: Initial draft

## Status
Proposed

## Context
We present some approaches on how we can slash a validator on the provider chain for
an equivocation performed on the consumer chain. Currently, we can receive [evidence of equivocation](https://github.com/cosmos/interchain-security/pull/1232), 
but we do not have functionality to slash the misbehaving validator on the provider chain.
In what follows, we first explain how slashing is performed on a single chain, to show why slashing on
the provider chain for a consumer equivocation is a challenging problem before proposing a potential solution.

### Single-chain slashing
Slashing is implemented across the [slashing](https://docs.cosmos.network/v0.50/modules/slashing)
and [staking](https://docs.cosmos.network/v0.50/modules/staking)

> modules. 

Author: Clarified this in the beginning (`v0.47.0` for Cosmos SDK and `v0.37` for CometBFT) and updated all referenced links to refer code from those specific versions.

this undelegation is **not** slashed, otherwise the undelegation is [slashed](https://github.com/cosmos/cosmos-sdk/blob/bf249162e42ef084668b942e12538cb30f9e4846/x/staking/keeper/slash.go#L262) by
`slashFactor`.

The slashing of redelegations happens in a similar way, meaning that `Slash` goes through all redelegations and checks on whether
the redelegation started before or after the `infractionHeight`.

We believe that one of the ideas behind using `infractionHeight` to decide on whether to slash an undelegation or redelegation
has to do with the fact that we want to slash only delegators whose voting power _contributed_ to the infraction.
However, this is a rather obscure idea because a delegator `D` could start unbonding at height `H` and then a validator
could intentionally perform an equivocation in the past (before `H`). In such a case, delegator `D` would still get slashed, even though `D`
did not contribute any voting power per se for the equivocation. Furthermore, this principle is not respected in general
(see "Slashing delegations" below).

#### Slashing delegations
Besides undelegations and redelegations, we need to slash simple delegations on the validator.

> This is performed by [deducting the appropriate amount of tokens](https://github.com/cosmos/cosmos-sdk/blob/5ca405ae067e7d8df98699f675e060f70a549976/x/staking/keeper/slash.go#L165)
> from the validator. Note that this deduction is computed based on the voting `power` the misbehaving validator

Reviewer 2: ```suggestion
This is performed by deducting the appropriate amount of tokens from the validator. Note that this deduction is computed based on the voting `power` the misbehaving validator
```

had at the height of the equivocation. As a result of the tokens deduction, 

> the [tokens per share](https://docs.cosmos.network/v0.46/modules/staking/01_state.html#delegator-shares)

Reviewer 2: This links to a different version of SDK.

reduce and hence later on, when delegators undelegate or redelegate, the delegators retrieve back less
tokens, effectively having their tokens slashed. This approach of slashing delegations does not utilize the

> `infractionHeight` in any way and hence the following could occur:

Reviewer 2: ```suggestion
`infractionHeight` in any way and hence the following scenario could occur:
```

  1. a validator `V` performs an equivocation at a height `Hi`
  2. a new delegator `D` delegates to `V` after height `Hi`

>   3. we receive the evidence of the equivocation by validator `V`

Reviewer 2: ```suggestion
  3. evidence of the equivocation by validator `V` is received
```

>   4. we slash the tokens of delegator `D`

Reviewer 2: ```suggestion
  4. the tokens of delegator `D` are slashed
```

In the above scenario, delegator `D` is slashed, even though `D`'s voting power did not contribute to the infraction. 

#### Old evidence

> In the single-chain case, we never act on old evidence (e.g., from 3 years ago). This is achieved through

Reviewer 2: ```suggestion
In the single-chain case, old evidence (e.g., from 3 years ago) is ignored. This is achieved through
```

> [CometBFT](https://docs.cometbft.com/v0.37/spec/consensus/evidence) that filters old evidence based on the parameters

Reviewer 2: This check is also done in the evidence module , see https://github.com/cosmos/cosmos-sdk/blob/eb853117a97cb07ec1b6d487ad24f7a14488a23e/x/evidence/keeper/infraction.go#L69

Author: Added a sentence that evidence is rechecked if it's old at the application (evidence module) level.

`MaxAgeNumBlocks` and `MaxAgeDuration` (see [here](https://github.com/cometbft/cometbft/blob/ae9826ed75ee411c0d809797ce209ee770c15c4f/evidence/pool.go#L266)).
In Cosmos Hub, the `MaxAgeNumBlocks` is set to 1000000 (i.e., ~70 days if we assume we need ~6 sec per block) and `MaxAgeDuration`

> is set to 172800000000000 ns (i.e., 2 days). Because of this check, we can easily exclude old evidence.

Reviewer 2: Are you sure about this value for `MaxAgeDuration`?

Author: I could not extract these values using `gaiad` (e.g., using something like `gaiad query evidence params`) so I downloaded the [genesis file](https://github.com/cosmos/mainnet/raw/master/genesis/genesis.cosmoshub-4.json.gz) found from [mintscan](https://www.mintscan.io/cosmos/parameters) and inside the file we have:
```
"evidence":
{
    "max_age_duration": "172800000000000",
    "max_age_num_blocks": "1000000",
    "max_bytes": "50000"
},
```
Looking at an arbitrary [block](https://www.mintscan.io/cosmos/blocks/16786240) we can see that BFT time is in nanosecond granularity (e.g., ` "timestamp": "2023-08-30T07:43:23.844982695Z"`) and I would expect `max_age_duration` to also be in nanoseconds because in the [code](https://github.com/cosmos/cosmos-sdk/blob/47f46643affd7ec7978329c42bac47275ac7e1cc/x/evidence/keeper/infraction.go#L45) we do:
```
ageDuration := ctx.BlockHeader().Time.Sub(infractionTime)
...
if ageDuration > cp.Evidence.MaxAgeDuration && ... 
```

Reviewer 4: This is how you do it using `gaiad`:
```
gaiad q params subspace baseapp EvidenceParams --node https://cosmos-rpc.polkachu.com:443

key: EvidenceParams
subspace: baseapp
value: '{"max_age_num_blocks":"1000000","max_age_duration":"172800000000000","max_bytes":"50000"}'
```

EvidenceParams are under `baseapp` subspace in v45
